compositor_render/src/transformations/shader/validation.rs

```diff
@@ -0,0 +1,416 @@
+
+    #[error("Sizes of an array don't match: {0:?} != {1:?}.")]
+    ArraySizeMismatch(ArraySizeOrConstant, ArraySizeOrConstant),
+
+    #[error("A composite type was used as an array length specifier.")]
+    // don't think this will ever happen
+    CompositeTypeAsArrayLen(ConstantInner),
+}
+
+#[derive(Debug)]
+pub enum ArraySizeOrConstant {
+    ArraySize(ArraySize),
+    Constant(Constant),
+}
+
+fn validate_type_equivalent(
+    ty1: Handle<Type>,
+    mod1: &Module,
+    ty2: Handle<Type>,
+    mod2: &Module,
+) -> Result<(), TypeEquivalenceError> {
+    let type1 = &mod1.types[ty1];
+    let type2 = &mod2.types[ty2];
+
+    if type1.name != type2.name {
+        return Err(TypeEquivalenceError::TypeNameMismatch(
+            type1.name.clone(),
+            type2.name.clone(),
+        ));
+    }
+
+    let ti1 = match type1.inner.canonical_form(&mod1.types) {
+        Some(t) => t,
+        None => type1.inner.clone(),
+    };
+    let ti2 = match type2.inner.canonical_form(&mod2.types) {
+        Some(t) => t,
+        None => type2.inner.clone(),
+    };
+
+    match ti1 {
+        naga::TypeInner::Scalar { .. }
+        | naga::TypeInner::Vector { .. }
+        | naga::TypeInner::Matrix { .. }
+        | naga::TypeInner::Atomic { .. }
+        | naga::TypeInner::Image { .. }
+        | naga::TypeInner::Sampler { .. }
+        | naga::TypeInner::AccelerationStructure
+        | naga::TypeInner::RayQuery
+        | naga::TypeInner::ValuePointer { .. } => {
+            if ti1 != ti2 {
+                return Err(TypeEquivalenceError::TypeStructureMismatch(
+                    type1.inner.clone(),
+                    type2.inner.clone(),
+                ));
+            }
+        }
+
+        naga::TypeInner::Array {
+            base: base1,
+            size: size1,
+            stride: stride1,
+        } => {
+            if let naga::TypeInner::Array {
+                base: base2,
+                size: size2,
+                stride: stride2,
+            } = ti2
+            {
+                if stride1 != stride2 {
+                    return Err(TypeEquivalenceError::TypeStructureMismatch(ti1, ti2));
+                }
+
+                validate_array_size_equivalent(size1, mod1, size2, mod2)?;
+                return validate_type_equivalent(base1, mod1, base2, mod2);
+            }
+            return Err(TypeEquivalenceError::TypeStructureMismatch(ti1, ti2));
+        }
+
+        naga::TypeInner::BindingArray {
+            base: base1,
+            size: size1,
+        } => {
+            if let naga::TypeInner::BindingArray {
+                base: base2,
+                size: size2,
+            } = ti2
+            {
+                validate_array_size_equivalent(size1, mod1, size2, mod2)?;
+                return validate_type_equivalent(base1, mod1, base2, mod2);
+            }
+            return Err(TypeEquivalenceError::TypeStructureMismatch(ti1, ti2));
```

Comment:
```suggestion
            let naga::TypeInner::BindingArray {
                base: base2,
                size: size2,
            } = ti2 else  {
                validate_array_size_equivalent(size1, mod1, size2, mod2)?;
                return validate_type_equivalent(base1, mod1, base2, mod2);
            }
            return Err(TypeEquivalenceError::TypeStructureMismatch(ti1, ti2));
```
I had to read this few times to realize this is a block for if statement. Maybe sth like this would be cleaner?